File: 24-course-schedule/20230312/course-schedule.js
'use strict';

const canFinish = function(numCourses, prerequisites) {
    // Build empty indegree list and adjacency list
    const indegree = [];
    const adjList = [];

    for (let i = 0; i < numCourses; ++i) {
        indegree[i] = 0;
        adjList[i] = [];
    }

    // Populate lists based on prerequisites array
    for (const [course, prereq] of prerequisites) {
        indegree[course]++;
        adjList[prereq].push(course);
    }

    // Create a stack of vertexes with indegree zero
    const noPreReqs = [];
    indegree.forEach((count, course) => count === 0 && noPreReqs.push(course));

    let sortedCount = 0;

    while (noPreReqs.length) {
        const zeroIndegreeCourse = noPreReqs.pop();
        const dependants = adjList[zeroIndegreeCourse];
        sortedCount++;

        // Decrement the number of indegrees for each dependant
        for (const dependant of dependants) {
            if (--indegree[dependant] === 0) {
                noPreReqs.push(dependant);
            }
        }
    }

    return sortedCount === numCourses;
}


module.exports = canFinish;

File: 24-course-schedule/20230312/course_schedule.py
from typing import List


class Solution:
    def canFinish(self, num_courses: int,
                  prerequisites: List[List[int]]) -> bool:
        adj_list = [[] for i in range(num_courses)]
        indegree = [0 for i in range(num_courses)]
        zero_stack = []
        sort_count = 0

        for course, prereq in prerequisites:
            indegree[course] += 1
            adj_list[prereq].append(course)

        for course in range(len(indegree)):
            if indegree[course] == 0:
                zero_stack.append(course)

        while zero_stack:
            current = zero_stack.pop()
            sort_count += 1
            connections = adj_list[current]

            for connection in connections:
                indegree[connection] -= 1

                if indegree[connection] == 0:
                    zero_stack.append(connection)

        return sort_count == num_courses
